### DISCUSSION TITLE
Help with webpack 4 and rendering a PDF via URL

### DISCUSSION THREAD

**Discussion by jdhorner** [wojtekmaj/react-pdf#1121] [Q&A] [answered]
Greetings everyone. I'll try and stay short and to the point. 

In **webpack.config.js**, this is our module section: 
_(note we're using webpack 4, and are stuck at that version for the moment)_ 
```js
  module: {
    rules: [
      {
        test: /node_modules\/pdfjs-dist.*\.js$/gi,  // this is the newly added object for
        use: ['file-loader']                        // the file-loader module
      },
      {
        test: /\.(ts|tsx|js|jsx|mdx)$/,
        exclude: /node_modules/,
        use: {
          loader: 'ts-loader',
          options: { transpileOnly: true }
        }
      },
      {
        test: /\.css$/,
        use: ['style-loader', 'css-loader']
      },
      {
        test: /\.(sass|scss)$/,
        use: ['style-loader', 'css-loader', 'sass-loader']
      },
      {
        test: /\.less$/,
        use: ['style-loader', 'css-loader', 'less-loader']
      },
      {
        test: /\.(jpeg|jpg|png|gif|svg|eot|ttf|woff|woff2)$/,
        use: [
          {
            loader: 'url-loader',
            options: { limit: 100000 }
          }
        ]
      }
    ]
  },
```

In my **PDFViewer.jsx** component (truncated for brevity) 
```js
import { Document, Page } from 'react-pdf/dist/esm/entry.webpack';

// note, URL is a fully-qualified pdf on a cdn; it requires no CORS of any kind
// and it works fine when pasted into an incognito window

<Document file={url} onLoadSuccess={onDocumentLoadSuccess}>
  <Page pageNumber={pageNumber} />
</Document>
```

But I end up getting this console error in Google Chrome: 

![image](https://user-images.githubusercontent.com/2692640/199601234-09ba09eb-e48b-4876-b5d8-b9fc0aee6f3a.png)

I have tried every combination of file-loader and url-loader that I can think of. But it seems nothing makes a difference. It's always the same error message. Obviously other local static files work fine, like images and fonts. But loading the pdf.worker.js file I just can't grasp. I feel like I must be missing something very obvious (and something I had **thought** importing from `'react-pdf/dist/esm/entry.webpack'` would solve. But I am wrong. 

I would love any help anyone can give me. All of the answers online seem to refer to loading local static PDFs.

**Comment by wojtekmaj** [comment]
I don't believe the file-loader rule for pdfjs-dist stuff is necessary. At least that's not what I do in Webpack 4 sample app.

**Reply by jdhorner** [reply]
OK; I've simplified it to be much more in line with your sample page for Webpack 4. I also switched it to be like your sample, where there is no file until I choose a local PDF from the file picker. It all works, but as soon as I choose a file, I get the same error: 

![image](https://user-images.githubusercontent.com/2692640/199741695-a00afd07-e8a6-4317-9c2e-7ff8640acf93.png)

There's a warning about setting up a fake worker, and then it complains about not being able to load file system files.

**Reply by jdhorner** [reply]
Here's the entire component now: 
```jsx
import { useState } from 'react';
import { Document, Page } from 'react-pdf/dist/esm/entry.webpack';

const PDFViewer = () => {
  const [file, setFile] = useState(null);
  const [numPages, setNumPages] = useState(null);

  function onFileChange(event) {
    setFile(event.target.files[0]);
  }

  function onDocumentLoadSuccess({ numPages: nextNumPages }) {
    setNumPages(nextNumPages);
  }

  return (
    <div className="Example">
      <header>
        <h1>react-pdf sample page</h1>
      </header>
      <div className="Example__container">
        <div className="Example__container__load">
          <label htmlFor="file">
            Load from file: <input onChange={onFileChange} type="file" />
          </label>
        </div>
        <div className="Example__container__document">
          <Document file={file} onLoadSuccess={onDocumentLoadSuccess}>
            {Array.from(new Array(numPages), (el, index) => (
              <Page key={`page_${index + 1}`} pageNumber={index + 1} />
            ))}
          </Document>
        </div>
      </div>
    </div>
  );
};

export default PDFViewer;
```

**Reply by jdhorner** [reply]
Oh my god. 

If I change line 3
`node_modules/react-pdf/dist/esm/entry.webpack.js`
from this: 
```js
import pdfjsWorker from 'file-loader!pdfjs-dist/legacy/build/pdf.worker';
```
to this
```js
import pdfjsWorker from 'url-loader!pdfjs-dist/legacy/build/pdf.worker';
```
it works!! But that seems hard-coded on your end. Is this something configurable??

**Reply by jdhorner** [reply]
I spoke too soon. It works for locally-selected files using your example. But switching it to use a URL as the file prop just gives a "failed to load PDF" message. (For the same PDF; [The URL is here](https://storage101.ord1.clouddrive.com/v1/MossoCloudFS_a867e4f9-37a9-4947-a315-3f3f3de03d29/DEV_private_2/a25b646a-9b94-48c5-b1dc-0d9a9e6bfb3a_Test_for_Josh_again.pdf?temp_url_sig=77fe125808b02435bd8a331aa8dcb071bd4b00d1&temp_url_expires=1667492821&inline). If I download that PDF and use it as a local file it works; if I use it as a URL for the file prop, it fails) :/

**Reply by wojtekmaj** [reply]
Hm. No ideas in my head just yet. I have no idea why loading a file from different location would affect worker URL.

Just a note that you're not bound to bundler-specific entries, and you can easily build one yourself if you need to change something.

**Reply by jdhorner** [reply, marked as answer]
OK. Well finally, I was able to get this to work by doing my own network request for the file as a blob, setting that blob result in state, and then passing that state to the <Document>'s file prop. Works like a charm. 

Sorry to have bothered you with all this. But it was helpful!
